### PR TITLE
2. Fetch  Streams

### PR DESCRIPTION
**Pull request by timrijckaert**
Addresses #21 

- [x] HLS streams
- [x] MPEG-DASH streams

### CODE REVIEW

File: goplay/src/main/java/be/tapped/goplay/model.kt [hunk 1]
```diff
@@ -39,6 +44,13 @@ public sealed interface ApiResponse {
             public object NoPrograms : Content
         }
 
+        public sealed interface Stream : Failure {
+            public data class MpegDash(val json: JsonObject, val throwable: Throwable) : Stream
```
Comment (timrijckaert) [inline goplay/src/main/java/be/tapped/goplay/model.kt:48]: Maybe I might condense this to 1 type. Since we can deduce from the JSON what went wrong

**Comment by timrijckaert** [comment]
I see 2 tests consistently failing:

- Don't Tell the Bride
  - S9E4
-  Hij is een Zij
   - S4E5

```
Network(throwable=kotlinx.serialization.json.internal.JsonDecodingException: Expected start of the object '{', but had 'EOF' instead
JSON input: "")
```
 
All the other tests seem to fail for the same reason 

```
Expected Either.Right, but found Either.Left with value Network(throwable=io.ktor.network.sockets.SocketTimeoutException: Socket timeout has expired [url=https://api.viervijfzes.be/content/38de3acc-86ef-4bc5-993a-c4ffbdfd8580, socket_timeout=unknown] ms)
```

This might be because we are almost DDOSing the GoPlay service in this test 🙈
We might need to do some backoff/retry (Arrow?) mechanisms.

File: goplay/src/main/java/be/tapped/goplay/model.kt (hunk 1 above)
Comment (timrijckaert) [reply goplay/src/main/java/be/tapped/goplay/model.kt:48]: Hmm or maybe not.

File: goplay/src/main/java/be/tapped/goplay/stream/StreamRepo.kt [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * Example of DRM response
+ *
+ * ```json
+ * {
+ *   "auth": "secret-authentication-token"
+ * }
+ * ```
+*/
+internal fun mpegDashStreamResolver(client: HttpClient): MpegDashStreamResolver =
```
Comment (timrijckaert) [inline goplay/src/main/java/be/tapped/goplay/stream/StreamRepo.kt:76]: Maybe the MpegDashStreamResolver should piggy back on the HlsStreamResolver if it fails.

**Review by nomisRev — APPROVED** [review]

**Comment by nomisRev** [comment]
> We might need to do some backoff/retry (Arrow?) mechanisms.

Resilience 🥳 I can make a small PR to add a utility for this, shall I?
You can find some info here: https://arrow-kt.io/docs/apidocs/arrow-fx-coroutines/arrow.fx.coroutines/-circuit-breaker/

**Comment by nomisRev** [comment]
Example of what we could do for resilience:

```kotlin
import arrow.core.Either
import arrow.fx.coroutines.Schedule.Companion.exponential
import arrow.fx.coroutines.Schedule.Companion.identity
import arrow.fx.coroutines.Schedule.Companion.recurs
import arrow.fx.coroutines.retry
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlin.time.ExperimentalTime
import arrow.fx.coroutines.CircuitBreaker as FxCircuitBreaker
import arrow.fx.coroutines.Schedule as FxSchedule

interface Resilience {
    suspend fun <A> resilient(action: suspend () -> A): A

    suspend fun <A> resilientCatch(action: suspend () -> A): Either<Throwable, A> =
        Either.catch { resilient(action) }
}

@OptIn(ExperimentalTime::class)
data class ResilientConfig(
    val circuitBreaker: CircuitBreaker = CircuitBreaker(),
    val schedule: Schedule = Schedule()
) {

    suspend fun toResilience(): Resilience {
        val breaker = toCircuitBreaker()
        return object : Resilience {
            override suspend fun <A> resilient(action: suspend () -> A): A =
                toSchedule<Throwable>().retry {
                    breaker.protectOrThrow(action)
                }
        }
    }

    private suspend fun toCircuitBreaker(): FxCircuitBreaker = FxCircuitBreaker.of(
        maxFailures = circuitBreaker.failureRateThreshold,
        resetTimeoutNanos = circuitBreaker.durationOfOpenState.inWholeNanoseconds.toDouble()
    )

    private fun <In> toSchedule(): FxSchedule<In, In> =
        (recurs<In>(schedule.maxRetries) and exponential(
            schedule.retryWaitDuration,
            schedule.retryBackoffMultiplier
        )) zipRight identity()

    data class CircuitBreaker(val failureRateThreshold: Int = 50, val durationOfOpenState: Duration = 4.seconds)

    data class Schedule(
        val retryWaitDuration: Duration = 500.milliseconds,
        val maxRetries: Int = 4,
        val retryBackoffMultiplier: Double = 2.5,
    )
}
```

File: goplay/src/main/java/be/tapped/goplay/stream/StreamRepo.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import io.ktor.client.HttpClient
+import io.ktor.client.request.headers
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.jsonObject
+import kotlinx.serialization.json.jsonPrimitive
+import org.apache.http.HttpHeaders
+
+internal fun interface StreamRepo {
+    suspend fun streamByVideoUuid(videoId: Program.Detail.Playlist.Episode.VideoUuid, idToken: IdToken): Either<ApiResponse.Failure, ApiResponse.Success.Stream>
```
[the remark this answers is not in the document]
Comment (timrijckaert) [reply goplay/src/main/java/be/tapped/goplay/stream/StreamRepo.kt:22]: Yeah indeed! 
We'll wait for it to land in stable 😎

File: goplay/src/main/java/be/tapped/goplay/stream/StreamRepo.kt [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+ * }
+ * ```
+*/
+internal fun mpegDashStreamResolver(client: HttpClient): MpegDashStreamResolver =
+    MpegDashStreamResolver { videoId, videoObj, idToken ->
+        either {
+            val mpegDash =
+                catch {
+                    val drmKey = videoObj.getValue("drmKey").jsonObject.getValue("S").jsonPrimitive.content
+                    val drmResponseJson = client.safeGet<JsonObject>("$apiGoPlay/video/xml/${drmKey}") { headers { append(HttpHeaders.AUTHORIZATION, idToken.token) } }.bind()
```
[the remark this answers is not in the document]
Comment (timrijckaert) [reply goplay/src/main/java/be/tapped/goplay/stream/StreamRepo.kt:82]: For now the only _authenticated_ calls we have to make are when we want to exchange a `VideoUuid` to a `Stream`.

However I could make a private extension function for this file?

**Pull request merged**